File: problems/dfs_template/attempt.cpp
Node<int>* dfs(Node<int>* root, int target) {
    if (root == nullptr) return nullptr;
    if (root -> val == target) return root;
    Node<int>* left = dfs(root-> left, target);
    if (left != nullptr) return left;
    Node<int>* right = dfs(root-> right, target);
    return right;
}

File: problems/valid_binary_search_tree/attempt.py
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

import sys
def valid_bst(root: Node) -> bool:
    """
    Traverse the tree with DFS, and check if the left val and right val are between for curr val, return max val
    """

# this function builds a tree from input; you don't have to modify it
# learn more about how trees are encoded in https://algo.monster/problems/serializing_tree
def build_tree(nodes, f):
    val = next(nodes)
    if val == "x":
        return None
    left = build_tree(nodes, f)
    right = build_tree(nodes, f)
    return Node(f(val), left, right)

if __name__ == "__main__":
    root = build_tree(iter(input().split()), int)
    res = valid_bst(root)
    print("true" if res else "false")
